### PR TITLE
fix(ext/node): Rewrite `node:v8` serialize/deserialize

### LINKED ISSUE #20613: Support Parcel
**Platform**

Darwin 22.6.0 arm64 arm

**Version**

deno 1.37.0

**Repro**

Make an index.html

```html
​<html>
  <head>
    <title>My First Parcel App</title>
  </head>
  <body>
    <h1>Hello, World!</h1>
  </body>
</html>
```

Run it with deno

- `deno run -A npm:parcel index.html`

**Expected**

```sh
Server running at http://localhost:1234
✨ Built in 494ms
```

**Actual**

```rust
➜ deno run -A npm:parcel index.html
Error: Not implemented: v8.deserialize
    at notImplemented (ext:deno_node/_utils.ts:9:9)
    at deserialize (node:v8:62:3)
    at deserialize (file:///Users/admin/repos/deno-parcel/node_modules/.deno/@parcel+core@2.9.3/node_modules/@parcel/core/lib/serializer.js:204:48)
    at loadRequestGraph (file:///Users/admin/repos/deno-parcel/node_modules/.deno/@parcel+core@2.9.3/node_modules/@parcel/core/lib/RequestTracker.js:744:52)
    at eventLoopTick (ext:core/01_core.js:183:11)
    at async Function.init (file:///Users/admin/repos/deno-parcel/node_modules/.deno/@parcel+core@2.9.3/node_modules/@parcel/core/lib/RequestTracker.js:718:17)
    at async Parcel._init (file:///Users/admin/repos/deno-parcel/node_modules/.deno/@parcel+core@2.9.3/node_modules/@parcel/core/lib/Parcel.js:178:28)
    at async Parcel.watch (file:///Users/admin/repos/deno-parcel/node_modules/.deno/@parcel+core@2.9.3/node_modules/@parcel/core/lib/Parcel.js:228:7)
    at async run (file:///Users/admin/repos/deno-parcel/node_modules/.deno/parcel@2.9.3/node_modules/parcel/lib/cli.js:286:9)
error: Uncaught (in worker "$DENO_STD_NODE_WORKER_THREAD") (in promise) Error: Only create ThreadsChild instances in a worker!
    at new ThreadsChild (file:///Users/admin/repos/deno-parcel/node_modules/.deno/@parcel+workers@2.9.3/node_modules/@parcel/workers/lib/threads/ThreadsChild.js:34:13)
    at new Child (file:///Users/admin/repos/deno-parcel/node_modules/.deno/@parcel+workers@2.9.3/node_modules/@parcel/workers/lib/child.js:65:18)
    at Object.<anonymous> (file:///Users/admin/repos/deno-parcel/node_modules/.deno/@parcel+workers@2.9.3/node_modules/@parcel/workers/lib/threads/ThreadsChild.js:49:27)
    at Object.<anonymous> (file:///Users/admin/repos/deno-parcel/node_modules/.deno/@parcel+workers@2.9.3/node_modules/@parcel/workers/lib/threads/ThreadsChild.js:50:4)
    at Module._compile (node:module:733:34)
    at Object.Module._extensions..js (node:module:747:10)
    at Module.load (node:module:658:32)
    at Function.Module._load (node:module:539:12)
    at Module.require (node:module:677:19)
    at require (node:module:791:16)

```

### PR DESCRIPTION
**Pull request by nathanwhit**
Closes #20613.

Reimplements the serialization on top of the v8 APIs instead of deno_core. Implements `v8.Serializer`, `v8.DefaultSerializer`, `v8.Deserializer`, and `v8.DefaultSerializer`.

### CODE REVIEW

File: ext/node/ops/v8.rs [hunk 1]
```diff
@@ -30,3 +39,343 @@ pub fn op_v8_get_heap_statistics(
...
+    &self,
+    scope: &mut v8::HandleScope<'s>,
+    message: v8::Local<'s, v8::String>,
+  ) {
+    let obj = self.obj(scope);
+    let key = FastString::from_static("_getDataCloneError")
+      .v8_string(scope)
+      .into();
+    if let Some(v) = obj.get(scope, key) {
+      if let Ok(fun) = v.try_cast::<v8::Function>() {
```
Comment (bartlomieju) [inline ext/node/ops/v8.rs:85]: Shouldn't this be an error/panic if it's not a function?

File: ext/node/ops/v8.rs [hunk 2]
```diff
@@ -30,3 +39,343 @@ pub fn op_v8_get_heap_statistics(
...
+  }
+}
+
+impl v8::ValueSerializerImpl for SerializerDelegate {
+  fn get_shared_array_buffer_id<'s>(
+    &self,
+    _scope: &mut v8::HandleScope<'s>,
+    _shared_array_buffer: v8::Local<'s, v8::SharedArrayBuffer>,
+  ) -> Option<u32> {
+    None
```
Comment (bartlomieju) [inline ext/node/ops/v8.rs:70]: Is that the final form or should this be fixed in the future?

File: ext/node/ops/v8.rs [hunk 3]
```diff
@@ -30,3 +39,343 @@ pub fn op_v8_get_heap_statistics(
...
+}
+
+#[op2(fast)]
+pub fn op_v8_transfer_array_buffer_de(
+  #[cppgc] deser: &Deserializer,
+  #[smi] id: u32,
+  array_buffer: v8::Local<v8::ArrayBuffer>,
+) {
+  // TODO(nathanwhit): also need binding for TransferSharedArrayBuffer, then call that if
+  // array_buffer is shared
```
Comment (bartlomieju) [inline ext/node/ops/v8.rs:301]: Is there an issue in `rusty_v8` repo for that?

File: ext/node/ops/v8.rs [hunk 4]
```diff
@@ -30,3 +39,343 @@ pub fn op_v8_get_heap_statistics(
...
+  #[cppgc] deser: &Deserializer,
+) -> Result<f64, AnyError> {
+  let mut double = 0f64;
+  if !deser.inner.read_double(&mut double) {
+    return Err(type_error("ReadDouble() failed"));
+  }
+  Ok(double)
+}
+
+#[op2(nofast)]
```
Comment (bartlomieju) [inline ext/node/ops/v8.rs:316]: Why does this particular op require `nofast` attr?

File: tests/node_compat/test/parallel/test-v8-serdes.js [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+//   des._readHostObject = common.mustCall(() => {
+//     const length = des.readUint32();
+//     const buf = des.readRawBytes(length);
+
+//     assert.strictEqual(buf.toString(), 'hostObjectTag');
+
+//     assert.deepStrictEqual(des.readUint64(), [1, 2]);
+//     assert.strictEqual(des.readDouble(), -0.25);
+//     return hostObject;
+//   });
```
Comment (bartlomieju) [inline tests/node_compat/test/parallel/test-v8-serdes.js:106]: Can you comment these tests why they are failing right now?

Comment (nathanwhit) [reply tests/node_compat/test/parallel/test-v8-serdes.js:106]: Oh they all rely on `hostObject` / JSStream (see the comment on line 42). I can add comments to that effect

File: ext/node/ops/v8.rs [hunk 6]
```diff
@@ -30,3 +39,343 @@ pub fn op_v8_get_heap_statistics(
   buffer[12] = stats.used_global_handles_size() as f64;
   buffer[13] = stats.external_memory() as f64;
 }
+
+pub struct Serializer<'a> {
+  inner: v8::ValueSerializer<'a>,
+}
+
+pub struct SerializerDelegate {
+  obj: v8::Global<v8::Object>,
```
Comment (devsnek) [inline ext/node/ops/v8.rs:48]: could this be a `v8::TracedReference`?

File: ext/node/ops/v8.rs [hunk 7]
```diff
@@ -30,3 +39,343 @@ pub fn op_v8_get_heap_statistics(
...
+
+pub struct Deserializer<'a> {
+  buf: DeserBuffer,
+  inner: v8::ValueDeserializer<'a>,
+}
+
+impl<'a> deno_core::GarbageCollected for Deserializer<'a> {}
+
+pub struct DeserializerDelegate {
+  obj: v8::Global<v8::Object>,
```
Comment (devsnek) [inline ext/node/ops/v8.rs:218]: same q

File: ext/node/ops/v8.rs (hunk 4 above)
Comment (nathanwhit) [reply ext/node/ops/v8.rs:316]: There's an op2 bug, see https://github.com/denoland/deno_core/issues/903

File: ext/node/ops/v8.rs (hunk 3 above)
Comment (nathanwhit) [reply ext/node/ops/v8.rs:301]: Opened https://github.com/denoland/rusty_v8/issues/1615

File: ext/node/ops/v8.rs (hunk 6 above)
Comment (nathanwhit) [reply ext/node/ops/v8.rs:48]: Maybe? I'm just not sure how to actually call `trace` on it. The delegate is owned by `v8::ValueSerializer`, which stores it as a `Box<dyn ValueSerializerImpl>` and doesn't know that it's garbage collected.

I guess I could maybe keep a `Rc<SerializerDelegate>` (or equivalent) around in `Serializer`, and trace that?

File: ext/node/ops/v8.rs (hunk 1 above)
Comment (nathanwhit) [reply ext/node/ops/v8.rs:85]: Ah yeah, it should panic (in node it's an assert)

File: ext/node/ops/v8.rs (hunk 2 above)
Comment (nathanwhit) [reply ext/node/ops/v8.rs:70]: In node this tries to call a function on the underlying JS object, but it's undocumented and unused, so I didn't bother originally.

Looking at it again though, it's super easy, so I'll just go ahead and match node's behavior here

File: ext/node/ops/v8.rs (hunk 6 above)
Comment (devsnek) [reply ext/node/ops/v8.rs:48]: oh i see i misread this since the gc impl came after SerializerDelegate, but is written for Serializer.

**Review by bartlomieju — APPROVED** [review]

**Pull request merged**